### ISSUE TITLE
Disk read io is high

### ISSUE THREAD

**Issue by canghai908** [m3db/m3#839]
I followed this tutorial   [https://m3db.github.io/m3/how_to/cluster_hard_way/](https://m3db.github.io/m3/how_to/cluster_hard_way/  )  to install and deploy three node clusters. After 10 minutes, the machine's io reached 200Mb/s, and the machine became very slow. Is this normal?
![image](https://user-images.githubusercontent.com/1450994/43990338-7bd5fa76-9d8d-11e8-9103-85499467c1ec.png)
The machine becomes very slow。Is there any parameter that can be optimized?
My m3dbnode-cluster.yml  is
```yaml
coordinator:
  listenAddress: 0.0.0.0:7201
  metrics:
    scope:
      prefix: "coordinator"
    prometheus:
      handlerPath: /metrics
      listenAddress: 0.0.0.0:7203 # until https://github.com/m3db/m3/issues/682 is resolved
    sanitization: prometheus
    samplingRate: 1.0
    extended: none

db:
  logging:
    level: info

  metrics:
    prometheus:
      handlerPath: /metrics
    sanitization: prometheus
    samplingRate: 1.0
    extended: detailed

  hostID:
    resolver: hostname

  config:
    service:
      env: default_env
      zone: embedded
      service: m3db
      cacheDir: /var/lib/m3kv
      etcdClusters:
        - zone: embedded
          endpoints:
            - 172.21.32.10:2379
            - 172.21.32.11:2379
            - 172.21.32.12:2379
    seedNodes:
      initialCluster:
        - hostID: m3db001.cactifans.com
          endpoint: http://172.21.32.10:2380
        - hostID: m3db002.cactifans.com
          endpoint: http://172.21.32.11:2380
        - hostID: m3db003.cactifans.com
          endpoint: http://172.21.32.12:2380

  listenAddress: 0.0.0.0:9000
  clusterListenAddress: 0.0.0.0:9001
  httpNodeListenAddress: 0.0.0.0:9002
  httpClusterListenAddress: 0.0.0.0:9003
  debugListenAddress: 0.0.0.0:9004

  client:
    writeConsistencyLevel: majority
    readConsistencyLevel: unstrict_majority
    writeTimeout: 10s
    fetchTimeout: 15s
    connectTimeout: 20s
    writeRetry:
        initialBackoff: 500ms
        backoffFactor: 3
        maxRetries: 2
        jitter: true
    fetchRetry:
        initialBackoff: 500ms
        backoffFactor: 2
        maxRetries: 3
        jitter: true
    backgroundHealthCheckFailLimit: 4
    backgroundHealthCheckFailThrottleFactor: 0.5

  gcPercentage: 100

  writeNewSeriesAsync: true
  writeNewSeriesLimitPerSecond: 1048576
  writeNewSeriesBackoffDuration: 2ms

  bootstrap:
    bootstrappers:
        - filesystem
        - peers
        - commitlog
        - noop-none
    fs:
        numProcessorsPerCPU: 0.125

  cache:
    series:
      policy: lru

  commitlog:
    flushMaxBytes: 524288
    flushEvery: 1s
    queue:
        calculationType: fixed
        size: 2097152
    retentionPeriod: 24h
    blockSize: 10m

  fs:
    filePathPrefix: /var/lib/m3db
    writeBufferSize: 65536
    dataReadBufferSize: 65536
    infoReadBufferSize: 128
    seekReadBufferSize: 4096
    throughputLimitMbps: 100.0
    throughputCheckEvery: 128

  repair:
    enabled: false
    interval: 2h
    offset: 30m
    jitter: 1h
    throttle: 2m
    checkInterval: 1m

  pooling:
    blockAllocSize: 16
    type: simple
    seriesPool:
        size: 262144
        lowWatermark: 0.7
        highWatermark: 1.0
    blockPool:
        size: 262144
        lowWatermark: 0.7
        highWatermark: 1.0
    encoderPool:
        size: 262144
        lowWatermark: 0.7
        highWatermark: 1.0
    closersPool:
        size: 104857
        lowWatermark: 0.7
        highWatermark: 1.0
    contextPool:
        size: 262144
        lowWatermark: 0.7
        highWatermark: 1.0
    segmentReaderPool:
        size: 16384
        lowWatermark: 0.7
        highWatermark: 1.0
    iteratorPool:
        size: 2048
        lowWatermark: 0.7
        highWatermark: 1.0
    fetchBlockMetadataResultsPool:
        size: 65536
        capacity: 32
        lowWatermark: 0.7
        highWatermark: 1.0
    fetchBlocksMetadataResultsPool:
        size: 32
        capacity: 4096
        lowWatermark: 0.7
        highWatermark: 1.0
    hostBlockMetadataSlicePool:
        size: 131072
        capacity: 3
        lowWatermark: 0.7
        highWatermark: 1.0
    blockMetadataPool:
        size: 65536
        lowWatermark: 0.7
        highWatermark: 1.0
    blockMetadataSlicePool:
        size: 65536
        capacity: 32
        lowWatermark: 0.7
        highWatermark: 1.0
    blocksMetadataPool:
        size: 65536
        lowWatermark: 0.7
        highWatermark: 1.0
    blocksMetadataSlicePool:
        size: 32
        capacity: 4096
        lowWatermark: 0.7
        highWatermark: 1.0
    identifierPool:
        size: 262144
        lowWatermark: 0.7
        highWatermark: 1.0
    bytesPool:
        buckets:
            - capacity: 16
              size: 524288
              lowWatermark: 0.7
              highWatermark: 1.0
            - capacity: 32
              size: 262144
              lowWatermark: 0.7
              highWatermark: 1.0
            - capacity: 64
              size: 131072
              lowWatermark: 0.7
              highWatermark: 1.0
            - capacity: 128
              size: 65536
              lowWatermark: 0.7
              highWatermark: 1.0
            - capacity: 256
              size: 65536
              lowWatermark: 0.7
              highWatermark: 1.0
            - capacity: 1440
              size: 16384
              lowWatermark: 0.7
              highWatermark: 1.0
            - capacity: 4096
              size: 8192
              lowWatermark: 0.7
              highWatermark: 1.0
```
This is the log of one of the nodes:

> [root@m3db002 m3db]# ./m3dbnode -f m3dbnode-cluster.yml 
2018/08/11 17:26:41 Go Runtime version: go1.10
2018/08/11 17:26:41 Build Version:      v0.4.0
2018/08/11 17:26:41 Build Revision:     8694e3e
2018/08/11 17:26:41 Build Branch:       master
2018/08/11 17:26:41 Build Date:         2018-08-10-11:46:38
2018/08/11 17:26:41 Build TimeUnix:     1533865598
17:26:41.040622[I] is a seed node; starting etcd server
2018-08-11T17:26:41.041+0800    INFO    constructed a logger
2018-08-11T17:26:41.041+0800    INFO    resolved cluster namespace      {"namespace": "default"}
2018-08-11T17:26:41.065+0800    INFO    starting server {"address": "0.0.0.0:7201"}
2018-08-11 17:26:41.080641 W | auth: simple token is not cryptographically signed
17:26:41.304813[W] max index query IDs concurrency was not set, falling back to default value
17:26:41.305461[W] host doesn't support HugeTLB, proceeding without it
17:26:41.328937[I] bytes pool registering bucket capacity=16, size=524288, refillLowWatermark=0.700000, refillHighWatermark=1.000000
17:26:41.328995[I] bytes pool registering bucket capacity=32, size=262144, refillLowWatermark=0.700000, refillHighWatermark=1.000000
17:26:41.329012[I] bytes pool registering bucket capacity=64, size=131072, refillLowWatermark=0.700000, refillHighWatermark=1.000000
17:26:41.329026[I] bytes pool registering bucket capacity=128, size=65536, refillLowWatermark=0.700000, refillHighWatermark=1.000000
17:26:41.329040[I] bytes pool registering bucket capacity=256, size=65536, refillLowWatermark=0.700000, refillHighWatermark=1.000000
17:26:41.329053[I] bytes pool registering bucket capacity=1440, size=16384, refillLowWatermark=0.700000, refillHighWatermark=1.000000
17:26:41.329075[I] bytes pool registering bucket capacity=4096, size=8192, refillLowWatermark=0.700000, refillHighWatermark=1.000000
17:26:41.329224[I] bytes pool simple init
2018-08-11 17:26:41.889817 W | etcdserver: failed to reach the peerURL(http://172.21.32.10:2380) of member f7ae4f836670b2a1 (Get http://172.21.32.10:2380/version: dial tcp 172.21.32.10:2380: connect: connection refused)
2018-08-11 17:26:41.889870 W | etcdserver: cannot get the version of member f7ae4f836670b2a1 (Get http://172.21.32.10:2380/version: dial tcp 172.21.32.10:2380: connect: connection refused)
17:26:43.893247[I] creating dynamic config service client with m3cluster
17:26:43.894987[I] successfully loaded cache from file /var/lib/m3kv/_kv_default_env_m3db_embedded.json
17:26:43.895052[I] adding a watch for service: m3db env: default_env zone: embedded includeUnhealthy: true
17:26:43.896400[I] successfully loaded cache from file /var/lib/m3kv/m3db_embedded.json
17:26:54.299680[I] cluster database initializing topology
17:26:54.299727[I] cluster database resolving topology
17:26:54.401245[I] resolving namespaces with namespace watch
17:26:54.401420[I] updating database namespaces [{adds [metrics]} {updates []} {removals []}]
17:26:54.876532[I] node tchannelthrift: listening on 0.0.0.0:9000
17:26:54.876920[I] cluster tchannelthrift: listening on 0.0.0.0:9001
17:26:55.300795[I] node httpjson: listening on 0.0.0.0:9002
17:26:55.301172[I] cluster httpjson: listening on 0.0.0.0:9003
17:27:06.266821[I] successfully updated topology to 3 hosts
17:27:07.751728[I] successfully updated topology to 3 hosts
17:27:08.026066[I] bootstrapping shards for range starting [{run bootstrap-data} {bootstrapper base} {namespace metrics} {numShards 1024} {from 2018-07-12 08:00:00 +0800 CST} {to 2018-08-11 08:00:00 +0800 CST} {range 720h0m0s}]
17:28:24.324014[I] bootstrapping from source starting [{source filesystem} {namespace metrics} {from 2018-07-12 08:00:00 +0800 CST} {to 2018-08-11 08:00:00 +0800 CST} {range 720h0m0s} {shards 1024}]
17:28:24.448378[I] filesystem bootstrapper resolving block retriever [{namespace metrics}]
17:28:24.448450[I] filesystem bootstrapper caching block retriever shard indices [{namespace metrics} {shards 1024}]
2018-08-11 17:28:37.390375 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 204.097269ms)
2018-08-11 17:28:37.390408 W | etcdserver: server is likely overloaded
2018-08-11 17:28:37.390419 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 204.180666ms)
2018-08-11 17:28:37.390431 W | etcdserver: server is likely overloaded
2018-08-11 17:28:45.233444 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 12.053491ms)
2018-08-11 17:28:45.233734 W | etcdserver: server is likely overloaded
2018-08-11 17:28:45.233851 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 26.524334ms)
2018-08-11 17:28:45.233898 W | etcdserver: server is likely overloaded
2018-08-11 17:28:46.539623 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 249.904473ms)
2018-08-11 17:28:46.568435 W | etcdserver: server is likely overloaded
2018-08-11 17:28:46.568469 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 591.551932ms)
2018-08-11 17:28:46.568483 W | etcdserver: server is likely overloaded
2018-08-11 17:28:46.568644 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 141.829915ms)
2018-08-11 17:28:46.568662 W | etcdserver: server is likely overloaded
2018-08-11 17:28:56.222691 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 38.474599ms)
2018-08-11 17:28:56.222773 W | etcdserver: server is likely overloaded
2018-08-11 17:28:56.222796 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 42.457388ms)
2018-08-11 17:28:56.222809 W | etcdserver: server is likely overloaded
2018-08-11 17:28:57.201440 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 25.005013ms)
2018-08-11 17:28:57.201493 W | etcdserver: server is likely overloaded
2018-08-11 17:28:57.201509 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 25.10294ms)
2018-08-11 17:28:57.201521 W | etcdserver: server is likely overloaded
2018-08-11 17:28:58.439779 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 45.776899ms)
2018-08-11 17:28:58.447520 W | etcdserver: server is likely overloaded
2018-08-11 17:28:58.447567 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 53.62846ms)
2018-08-11 17:28:58.447581 W | etcdserver: server is likely overloaded
2018-08-11 17:29:00.320901 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 233.997699ms)
2018-08-11 17:29:00.348483 W | etcdserver: server is likely overloaded
2018-08-11 17:29:00.348512 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 272.169606ms)
2018-08-11 17:29:00.348525 W | etcdserver: server is likely overloaded
2018-08-11 17:29:01.794495 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 487.685888ms)
2018-08-11 17:29:01.794584 W | etcdserver: server is likely overloaded
2018-08-11 17:29:01.794603 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 487.837315ms)
2018-08-11 17:29:01.794616 W | etcdserver: server is likely overloaded
2018-08-11 17:29:02.011599 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 17.143906ms)
2018-08-11 17:29:02.011640 W | etcdserver: server is likely overloaded
2018-08-11 17:29:02.011658 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 17.055498ms)
2018-08-11 17:29:02.011671 W | etcdserver: server is likely overloaded
2018-08-11 17:29:02.216357 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 4.683479ms)
2018-08-11 17:29:02.216407 W | etcdserver: server is likely overloaded
2018-08-11 17:29:02.216424 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 4.838076ms)
2018-08-11 17:29:02.216436 W | etcdserver: server is likely overloaded
2018-08-11 17:29:09.265244 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 381.713371ms)
2018-08-11 17:29:09.273344 W | etcdserver: server is likely overloaded
2018-08-11 17:29:09.273370 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 389.885348ms)
2018-08-11 17:29:09.273382 W | etcdserver: server is likely overloaded
2018-08-11 17:29:17.211829 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 657.310347ms)
2018-08-11 17:29:17.211934 W | etcdserver: server is likely overloaded
2018-08-11 17:29:17.211957 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 657.493055ms)
2018-08-11 17:29:17.211981 W | etcdserver: server is likely overloaded
2018-08-11 17:29:21.217591 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 41.1563ms)
2018-08-11 17:29:21.217667 W | etcdserver: server is likely overloaded
2018-08-11 17:29:21.217687 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 41.326031ms)
2018-08-11 17:29:21.217700 W | etcdserver: server is likely overloaded
2018-08-11 17:29:22.126209 W | rafthttp: lost the TCP streaming connection with peer f7ae4f836670b2a1 (stream MsgApp v2 reader)
2018-08-11 17:29:22.130244 W | rafthttp: lost the TCP streaming connection with peer f7ae4f836670b2a1 (stream Message reader)
2018-08-11 17:29:22.227324 E | rafthttp: failed to dial f7ae4f836670b2a1 on stream MsgApp v2 (dial tcp 172.21.32.10:2380: connect: connection refused)
2018-08-11 17:29:22.387855 W | rafthttp: lost the TCP streaming connection with peer f7ae4f836670b2a1 (stream Message writer)
2018-08-11 17:29:22.888513 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 109.200616ms)
2018-08-11 17:29:22.888579 W | etcdserver: server is likely overloaded
2018-08-11 17:29:22.888601 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 112.093998ms)
2018-08-11 17:29:22.888614 W | etcdserver: server is likely overloaded

 My environment is configured as follows
Hardware: 2vcpu/4G DDR4/20G SSD
Os:centos7.5_x86_64

**Comment by robskillington** [comment]
Hey @canghai908, thanks a lot for submitting the issue with a lot of detail.  We're very keen on helping fix these "teething" issues within a short amount of time, so apologies for not responding earlier.

My theory is that because each node has ownership of the whole shard range, that a long retention is likely very difficult as there are a ton of file descriptors its trying to create.

Could you please help list what retention and block size you configured?  The guide specifies 720h (which is 30d) and a block size of 12h.  This would create 7 files (per file volume) multiplied by num shards 1024 multiplied by num blocks 60 (retention 720h / block size 12h) which will result in 430,080 file descriptors.  This isn't too insane but it's definitely more than we run with in production (usually less shards per node).  There's been some discussion amongst about combining the fileset volumes across the shards, however this will complicate the removal of unowned shards.  It would greatly reduce the number of file descriptors needed however and all this IO seeking the directories and file volumes.

Anyway, let me reproduce the problem and explore the different options to fix it, so we can see if there's not a pragmatic quick fix.

**Comment by robskillington** [comment]
Hey, just an update, I was able to reproduce this.  I'm working through it, it actually doesn't seem like the issue I described and is likely a more minor bug.

I'm using the following:
- 1 vCPU
- 3.75GB memory
- 20gb non-SSD
- Centos 7

It may be related to the bootstrappers configuration.  I'll post more in the next 24 hours.

**Comment by canghai908** [comment]
Thanke you.I will continue to pay attention to this issue.

**Comment by robskillington** [comment]
@canghai908 can you detail which retention and block size you were using?  Those from the guide or something other?

**Comment by canghai908** [comment]
I use 
```
curl -X POST localhost:7201/api/v1/namespace -d '{
  "name": "metrics",
  "options": {
    "bootstrapEnabled": true,
    "flushEnabled": true,
    "writesToCommitLog": true,
    "cleanupEnabled": true,
    "snapshotEnabled": false,
    "repairEnabled": false,
    "retentionOptions": {
      "retentionPeriodDuration": "720h",
      "blockSizeDuration": "12h",
      "bufferFutureDuration": "1h",
      "bufferPastDuration": "1h",
      "blockDataExpiry": true,
      "blockDataExpiryAfterNotAccessPeriodDuration": "5m"
    },
    "indexOptions": {
      "enabled": true,
      "blockSizeDuration": "12h"
    }
  }
}'
```
to create a namespace.

**Comment by robskillington** [comment]
@canghai908 can you try again with a more suitable shard count for a three node cluster on small VMs?  Using say `"num_shards": 32`?

I'm seeing if there's anything we can do so you don't need to size this up front so much, but that's likely a good solution to pick roughly the amount of shards you may need up front based on the size of workload you'll be using the cluster for roughly.

**Comment by canghai908** [comment]
@robskillington  Thank you! I used` "num_shards": 32 `  the host's io is ok！The cluster is ok!

**Comment by richardartoul** [comment]
@canghai908 There are some other improvements we still need to make, but we just landed #853 which should help with the issue you experienced.
